packages/vats/src/distributeFees.js

```diff
@@ -0,0 +1,108 @@
+// @ts-check
+
+import { observeNotifier } from '@agoric/notifier';
+import { E } from '@agoric/eventual-send';
+import { natSafeMath } from '@agoric/zoe/src/contractSupport';
+import { amountMath } from '@agoric/ertp';
+import { Far } from '@agoric/marshal';
+
+const { add, subtract, multiply, floorDivide } = natSafeMath;
+
+// wrapper to take the treasury's creatorFacet, and make a function that will
+// request an invitation and return a promise for a payment.
+export function makeTreasuryFeeCollector(zoe, treasuryCreatorFacet) {
+  /** @type FeeCollector */
+  return Far('collectFees', {
+    collectFees: () => {
+      const invitation = E(treasuryCreatorFacet).makeCollectFeesInvitation();
+      const collectFeesSeat = zoe.offer(invitation, undefined, undefined);
+      return E(collectFeesSeat).getPayout('RUN');
+    },
+  });
+}
+
+// A distributor of fees from treasury to the Bank module, which has a mapping
+// from accounts to purses. Each time the epochTimer signals the end of an
+// Epoch, it will ask the bank's notifier for a fresh list of accounts and ask
+// the treasury for the fees that have been collected to date. It will then
+// divide the funds evenly amount the accounts, and send payments in batches of
+// depositsPerUpdate every updateInterval. When the payment doesn't divide
+// evenly among the accounts, it holds the remainder till the next epoch.
+
+/** @type {BuildFeeDistributor} */
+export function buildDistributor(treasury, bank, epochTimer, timer, params) {
+  const {
+    depositsPerUpdate,
+    updateInterval,
+    epochInterval,
+    runIssuer,
+    runBrand,
+  } = params;
+  const accountsNotifier = bank.getAccountsNotifier();
+  let leftOverPayment;
+  let leftOverValue = 0n;
+  const queuedAccounts = [];
+  const queuedPayments = [];
+  let lastWallTimeUpdate;
+  const timerNotifier = timer.makeNotifier(0n, updateInterval);
+
+  async function scheduleDeposits() {
+    if (!queuedPayments.length) {
+      return;
+    }
+
+    ({ updateCount: lastWallTimeUpdate } = await timerNotifier.getUpdateSince(
+      lastWallTimeUpdate,
+    ));
+
+    bank.depositMultiple(
```

Comment:
Since this is after an `await`, we should recheck the invariant:
```suggestion
    // queuedPayments may have changed since the `await`.
    if (!queuedPayments.length) {
      return;
    }
    bank.depositMultiple(
```

Reply:
Thanks. I was trying to think through the race paths and missed this one.